
# METASOFT pipeline

# config.yaml
# datasets: 
# - dataset1
# - dataset2
# outFolder: "."
# dataCode: "test"
import os
datasets = config["datasets"]
chromosomes = config["chrom"]

#prefix = "-".join(datasets)
prefix = "merged_cohorts"

outFolder = config["outFolder"]

dataset_string = os.path.join(outFolder, prefix)
print(dataset_string)

#chromosomes = config["chrom"]
print(chromosomes)

if chromosomes == "all": 
    chromosomes = list(range(1,23))

print("chromosomes = ", chromosomes)
#chromosomes = list(range(1,23))
vcf = config["vcf"]

rule all:
    input:
        expand(dataset_string + ".metasoft.{filetype}.qval.tsv.gz",  filetype = ["top_assoc"] ),
	expand(dataset_string + ".metasoft.{filetype}.tabixed.tsv.gz.tbi", filetype = "full_assoc" )
	#expand(dataset_string + ".{tool}.{filetype}.qval.tsv.gz",  tool = ["metasoft", "re2c"], filetype = ["top_assoc"] ),
        #expand(dataset_string + ".{tool}.{filetype}.tabixed.tsv.gz.tbi", tool = ["metasoft", "re2c"], filetype = "full_assoc" )


rule createInputFiles:
    output:
       dataset_string + ".{chr}.input.tsv"
    params:
        dataset_args = " ".join(datasets)
    shell:
        "ml R/4.0.3;"
        "Rscript scripts/createMetaInputByChr.R -c {wildcards.chr} -o {outFolder} {params.dataset_args} "


rule runMETASOFT:
    input:  dataset_string + ".{chr}.input.tsv"
    output: dataset_string + ".{chr}.metasoft.res.tsv"
    params:
        jarfile = "/sc/arion/projects/ad-omics/data/software/METASOFT/Metasoft.jar",
        ptable = "/sc/arion/projects/ad-omics/data/software/METASOFT/HanEskinPvalueTable.txt",
        logfile = dataset_string + ".{chr}.metasoft.log.txt" 
    shell:
        "ml java; java -jar {params.jarfile} -pvalue_table {params.ptable} -input {input} -mvalue -mvalue_p_thres 1E-4 -output {output} -log {params.logfile} -verbose"

rule runRE2C:
    input: dataset_string + ".{chr}.input.tsv"
    output: dataset_string + ".{chr}.re2c.res.tsv"
    params:
        re2c = "/sc/arion/projects/ad-omics/data/software/RE2Cv1_04/RE2C.sh"
    shell:
        "ml R/4.0.3;"
        "sh {params.re2c} --input {input} --output {output}; "
        "mv {output}.txt {output}" # RE2C appends .txt to end of output - remove it

rule runRECOV:
    input: dataset_string + ".{chr}.input.tsv"
    output: dataset_string + ".{chr}.recov.res.tsv"
    params:
        RECOV = "/sc/arion/projects/ad-omics/data/software/RECOV/fitLlh1Gene4github.R",
        logfile = dataset_string + ".{chr}.recov.log.txt"
    shell:
        "ml R/4.0.3;"
        #"Rscript {params.RECOV} {input} {output}" 
        "R CMD BATCH --no-save --no-restore \"--args {input} {output} 10 \" {params.RECOV} {params.logfile}"

rule processVCF:
    input: vcf
    output: outFolder + "vcf.processed.txt"
    shell:
        "ml bcftools; bcftools query -f '%CHROM\t%POS\t%ID\n'  {input} > {output} " 

# split phenotype from SNP, add coords from VCF
rule formatResults:
    input: 
        res = dataset_string + ".{chr}.{tool}.res.tsv",
        vcf_df = outFolder + "vcf.processed.txt"
    output:  
        full = dataset_string + ".{chr}.{tool}.full_assoc.tsv",
        top = dataset_string + ".{chr}.{tool}.top_assoc.tsv"
    shell:
        "ml R/4.0.3;"
        "Rscript scripts/formatMETASOFT.R -c {wildcards.chr} -i {input.res} -o {dataset_string}.{wildcards.chr}.{wildcards.tool} -v {input.vcf_df} "

# merge together and tabix
rule mergeFullResults:
    input: 
        expand( dataset_string + ".{chr}.{tool}.{filetype}.tsv", chr = chromosomes, filetype = "full_assoc", allow_missing=True ),
    output: 
        gzip = dataset_string + ".{tool}.{filetype}.tabixed.tsv.gz",
        tbi = dataset_string + ".{tool}.{filetype}.tabixed.tsv.gz.tbi"
    shell:
        "ml bcftools/1.9;"
        "cat {input} | bgzip > {output.gzip};"
        "tabix -s 3 -b 4 -e 4 -S 1 {output.gzip}"     

## bring together top association per feature
## run qvalue to correct for number of features tested
## no tabix because files are sorted by 
rule mergeTopResults:
    input:
        expand( dataset_string + ".{chr}.{tool}.{filetype}.tsv", chr = chromosomes, filetype = "top_assoc", allow_missing=True ) 
    output:
        gzip = dataset_string + ".{tool}.top_assoc.qval.tsv.gz"
    params:
        tmp = dataset_string + ".{tool}.top_assoc.tmp.tsv.gz"
    shell:
        "ml R/4.0.3; ml bcftools/1.9;"
        "cat {input} | gzip > {params.tmp};"
        "Rscript scripts/qvalueMETASOFT.R -i {params.tmp} -o {output}"
