# TWAS

Transcriptome-wide Association Studies

Requires:

* VCF
* phenotype BED file (generated by QTL-mapping pipeline)
* covariates file (generated by QTL-mapping pipeline)

* GWAS summary stats - must be munged by LDSC


## Standard FUSION TWAS

Steps:

1) Build TWAS weights (create_fusion_weights.smk)

2) Run TWAS association with GWAS summary stats (run_fusion_twas.smk)

## FOCUS fine-mapping

This assumes you have multiple TWAS panels. 

Both steps are implemented by run_focus.smk

1) Combine TWAS panels together into a FOCUS database

2) Run FOCUS fine-mapping



